contrib/graphite/blueflood.py

```diff
@@ -87,105 +217,207 @@ def fetch(self, start_time, end_time):
+  def __init__(self, host, tenant, enable_submetrics, submetric_aliases):
     self.host = host
     self.tenant = tenant
+    self.enable_submetrics = enable_submetrics
+    self.submetric_aliases = submetric_aliases
+    # This is the maximum number of json characters permitted by the
+    #  BF multiplot handler.
+    # The actual max is 8000, but I want to leave a margin for error
+    #  because I can't be certain of the exact way the json is generated
+    self.maxlen_per_req = 7000
+    # this is for: "'' ," that surround each metric
+    self.overhead_per_metric = 4
+    self.maxmetrics_per_req = 100
 
-  def make_request(self, url, payload, headers):
+  def gen_data_key(self, values):
+    # Determines which key to use for the data
+    if not len(values):
+      return None
+    value_res_order = ['average', 'latest', 'numPoints']
+    present_keys = [v for v in value_res_order if v in values[0]]
+    if present_keys:
+      return present_keys[0]
+    else:
+      return None
+
+  def current_datapoint_passed(self, v_iter, ts):
+    # Determines if the current datapoint is too old to be considered
+    if not len(v_iter):
+      return False
+    datapoint_ts = v_iter[0]['timestamp']/1000
+    if (ts > datapoint_ts):
+      return True
+    return False
+
+  def current_datapoint_valid(self, v_iter, data_key, ts, step):
+    # Determines if the current datapoint be included in the current step
+    #assumes current_datapoint_passed() is true
+    if (not len(v_iter)) or not (data_key in v_iter[0]):
+      return False
+    datapoint_ts = v_iter[0]['timestamp']/1000
+    if (datapoint_ts < (ts + step)):
+      return True
+    return False
+
+  def process_path(self, values, start_time, end_time, step, data_key):
+    # Generates datapoints in graphite-api format
+    # Graphite-api requires the points be "step" seconds apart in time
+    #  with Null's interleaved if needed
+    if not data_key:
+      print "No valid keys present"
+      return []
+    v_iter = values
+    ret_arr = []
+    for ts in range(start_time, end_time, step):
+      # Skip datapoints that have already passed
+      #  NOTE/TODO: this while loop has the effect of dropping all but the first datapoint in each step
+      #  (Graphite requires exactly one datapoint/step.)  It would be better to rollup the datapoints
+      #  if there are more than one/step.  However that will only happen when we have full res metrics
+      #  with sub-second frequencies.  We believe that case is too rare to worry about right now.
+      while self.current_datapoint_passed(v_iter, ts):
+        v_iter = v_iter[1:]
+      if self.current_datapoint_valid(v_iter, data_key, ts, step):
+        ret_arr.append(v_iter[0][data_key])
+      else:
+        ret_arr.append(None)
+
+    return ret_arr
+
+  def get_multi_endpoint(self, endpoint, tenant):
+    return "%s/v2.0/%s/views" % (endpoint, tenant)
+
+  def get_metric_list(self, endpoint, tenant, metric_list, payload, headers):
```

Comment:
This name of the function might be a bit confusing, as it prompts to metrics_list api call. Can we have a better name of it, may be `get_metric_data` or something like that?